pallets/settlement/src/lib.rs

```diff
@@ -645,7 +652,11 @@ decl_module! {
             let legs = InstructionLegs::iter_prefix(instruction_id).collect::<Vec<_>>();
 
             // Ensure that the sender is a party of this instruction.
-            ensure!(legs.iter().any(|(_, leg)| Self::is_leg_party(leg, primary_did, secondary_key.as_ref())), Error::<T>::UnauthorizedSigner);
+            T::Portfolio::ensure_portfolio_custody_and_permission(portfolio, primary_did, secondary_key.as_ref())?;
+            ensure!(
+                legs.iter().take(max_legs_count as usize).any(|(_, leg)| leg.from == portfolio || leg.to == portfolio),
+                Error::<T>::UnauthorizedSigner
+            );
```

Comment:
`legs.iter().take(max_legs_count)` will only take the first `max_legs_count`, items from the iterator.

I would recommend:
```
ensure!(legs.iter().filter(|(_, leg)| leg.from == portfolio || leg.to == portfolio)
  .count() <= max_legs_count, Error::<T>::LegCountTooSmall
);
```
We need to ensure that `max_legs_count` is not smaller than the number of legs that matches the portfolio.  Otherwise the user can always pass 1 to save on transaction costs.

Reply:
@Neopallium A user can only pass 1 and successfully reject the instruction if they are involved in the first leg. That seems to be a correct approach as it will save the chain compute. If you pass 1 and you aren't in that first leg than the check will fail early. I don't know how this can be abused and provides a benefit for users that know what leg they are involved in.

Reply:
It wouldn't save any compute, even if the user passes 10 and their portfolio is in the first leg, the `any()` will return `true` after checking the first leg.  `iter().take(10)` is not going to consume 10 items, it is more a limit on how many items can be taken from the original iterator.